### PR TITLE
Update Scoop support

### LINKED ISSUE #401: Scoop app detection has been broken
Scoop has [changed](https://github.com/ScoopInstaller/Scoop/commit/c4d1b9c22f943a810bed7f9f74d7d4d5c42d9a74#diff-d83ac0b4261ffb627685894a5ad95600470725a984e03c810b2cc3e45f5495a0) its export format to “pretty” JSON and the app detection is now failing. These changes were done on June 24 and released on July 7. I can help with updating [ScoopFactory.cs](https://github.com/Klocman/Bulk-Crap-Uninstaller/blob/4494080365f8e41673cf1c6759847adefd1f1569/source/UninstallTools/Factory/ScoopFactory.cs), but I wonder if it is necessary to retain compatibility with old format.

### PR DESCRIPTION
**Pull request by davidxuang**
Closes #401 

---

- [x] Parse JSON export format
  - [x] Custom `DateTimeOffset` converter for Windows PowerShell compatibility
  - [x] Set `InstallDate` via `Updated`
  - [ ] ~~Set `InstallSource` via `Source`~~
  - [ ] ~~Set `RatingId` via `Source`~~
- [x] Parse app install and manifest JSON file
  - [x] Custom `string[]` converter for `bin` and `env_add_path`
  - [x] Set `AboutUrl` via `homepage`
  - [x] Set `SortedExecutables` via `shortcuts`, `bin`, and `env_add_path`
    - [x] `architecture` awareness
  - [x] Set `IconBitmap` via `shortcuts` with explicit ico file
- [x] `JsonSerializerContext` support for accelerating JSON parsing

### CODE REVIEW

**Comment by davidxuang** [comment]
Should be working now, but not sure if bucket entry should be parsed (for `InstallSource` or `RatingId` maybe). Also the code is still a bit messy.

**Comment by davidxuang** [comment]
> InstallSource should be a path, so using value of `Source` won't work as it is. RatingId should end up being the same as it was in the old version so that user ratings are not lost.

`Source` could be `$env:SCOOP\buckets\$($bucket)\bucket`, though it's a bit strange if file manager don't scroll to the specific JSON file, since there isn't such property as `InstallerName`.

File: source/UninstallTools/Factory/ScoopFactory.cs [hunk 1]
```diff
@@ -92,59 +126,172 @@ public IList<ApplicationUninstallerEntry> GetUninstallerEntries(ListGenerationPr
...
+                if (executables.Any())
                 {
-                    RawDisplayName = name,
-                    DisplayVersion = ApplicationEntryTools.CleanupDisplayVersion(version),
-                    RatingId = "Scoop " + name
-                };
-
-                var installDir = Path.Combine(isGlobal ? _scoopGlobalPath : _scoopUserPath, "apps\\" + name);
-                if (Directory.Exists(installDir))
+                    entry.SortedExecutables = executables.Distinct().ToArray();
```
Comment (Klocman) [inline source/UninstallTools/Factory/ScoopFactory.cs:279]: SortedExecutables should be sorted or some unexpected things could happen later.
```suggestion
                    entry.SortedExecutables = AppExecutablesSearcher.SortListExecutables(
                            executables.Distinct().Select(x => new FileInfo(x)),
                            entry.DisplayNameTrimmed)
                        .Select(x => x.FullName)
                        .ToArray();
```

**Review by Klocman — COMMENTED** [review]
Everything looks to be working correctly, I only noticed one small issue.

> `Source` could be `$env:SCOOP\buckets\$($bucket)\bucket`, though it's a bit strange if file manager don't scroll to the specific JSON file, since there isn't such property as `InstallerName`.

InstallSource can be left empty in this case. RatingId stays the same. In that case the PR looks good to merge after you comment on this suggestion (or merge it as it is).

**Comment by davidxuang** [comment]
> InstallSource can be left empty in this case. RatingId stays the same. In that case the PR looks good to merge after you comment on this suggestion (or apply it as it is).

This is fine since open the source JSON file doesn't seems helpful in most cases.

> Also, what do you think about #388? `--purge` could be added to all Scoop uninstall commands.

I'd say "persist" should be deleted only when explicitly opted in. Also the persist are quite easy to locate, so maybe just add it to clean-up candidates.

**Comment by davidxuang** [comment]
Ahh, I notice that there might be a typo/misnaming. (`isPublic` should be `isGlobal`)

**Comment by Klocman** [comment]
It doesn't really matter, in this context it means the same thing.

**Pull request merged**